src/mrinufft/trajectories/tools.py

```diff
@@ -369,6 +380,290 @@ def unepify(trajectory: NDArray, Ns_readouts: int, Ns_transitions: int) -> NDArr
+    """Calculate the gi value for the trapezoidal gradient waveform."""
+    return (2 * area_needed - (n_down + 1) * gs - (n_up - 1) * ge) / (
+        n_down + n_up + 2 * n_pl
+    )
+
+
+def get_gradient_times_to_travel(
+    kspace_end_loc: Optional[NDArray] = None,
+    kspace_start_loc: Optional[NDArray] = None,
+    end_gradients: Optional[NDArray] = None,
+    start_gradients: Optional[NDArray] = None,
+    gamma: float = Gammas.Hydrogen,
+    raster_time: float = DEFAULT_RASTER_TIME,
+    gmax: float = DEFAULT_GMAX,
+    smax: float = DEFAULT_SMAX,
+    n_jobs: int = 1,
+) -> tuple[NDArray, NDArray, NDArray, NDArray]:
+    """Get gradient timing values for trapezoidal or triangular waveforms.
+
+    Compute gradient timing values to take k-space trajectories
+    from position ``ks`` with gradient ``gs`` to position ``ke`` with gradient
+    ``ge``, while being hardware compliant.
+    This function calculates the minimal number of time steps required for
+    the ramp down, ramp up, and plateau phases of the gradient waveform,
+    ensuring that the area traversed in k-space matches the desired
+    trajectory while adhering to the maximum gradient amplitude and
+    slew rate constraints.
+
+    Parameters
+    ----------
+    kspace_end_loc : NDArray
+        Ending k-space positions, shape (nb_shots, nb_dimension).
+    kspace_start_loc : NDArray, default None when it is 0
+        Starting k-space positions, shape (nb_shots, nb_dimension).
+    end_gradients : NDArray, default None when it is 0
+        Ending gradient values, shape (nb_shots, nb_dimension).
+    start_gradients : NDArray, default None when it is 0
+        Starting gradient values, shape (nb_shots, nb_dimension).
+    gamma : float, optional
+        Gyromagnetic ratio in Hz/T. Default is Gammas.Hydrogen.
+    raster_time : float, optional
+        Time interval between gradient samples (s). Default is DEFAULT_RASTER_TIME.
+    gmax : float, optional
+        Maximum gradient amplitude (T/m). Default is DEFAULT_GMAX.
+    smax : float, optional
+        Maximum slew rate ``T/m/s``. Default is DEFAULT_SMAX.
+    n_jobs : int, optional
+        Number of parallel jobs to run for optimization, by default 1.
+
+    Returns
+    -------
+    n_ramp_down: The timing values for the ramp down phase.
+    n_ramp_up: The timing values for the ramp up phase.
+    n_plateau: The timing values for the plateau phase.
+    gi: The intermediate gradient values for trapezoidal or triangular waveforms.
+    """
+    area_needed = (kspace_end_loc - kspace_start_loc) / gamma / raster_time
+
+    def solve_gi_min_plateau(gs, ge, area):
+        def _residual(gi):
+            n_down, n_up = _calculate_ramps(gs, ge, gi, smax, raster_time)
+            n_pl = _calculate_plateau(gs, ge, gi, n_down, n_up, area)
+            if n_pl < 0:
+                return np.abs(n_pl) * 10000  # Penalize negative plateau
+            return n_pl * 100
```

Comment:
Why the 100 factor ? 
Also why are negative plateau a bad idea ?

Reply:
It is not negative plateau, but rather negative discrete time events during which plateau occurs. By definition it should be greater than 0.
In fact all `n_*` needs to be >0 and integers. (n_ramp_up, n_ramp_down, n_pl)